lib/puppet/util/windows/adsi.rb

```diff
@@ -94,6 +111,14 @@ def execquery(query)
       wmi_connection.execquery(query)
     end
 
+    def domain_role
+      unless @domain_role
+        query_result = Puppet::Util::Windows::ADSI.execquery('select DomainRole from Win32_ComputerSystem').to_enum.first
+        @domain_role = DOMAIN_ROLES[query_result.DomainRole] if query_result
```

Comment:
For reference in case anyone comes back looking at this PR.

While it is possible that `to_enum` could return an empty array, the likelihood is VERY small, and would indicate WMI is broken or the API contract in WMI has changed in a breaking way.

It's more likely that a new item would be added to the enumeration that this PR doesn't cater for, however right now that would have no impact, and would just cause the memo-ization to not work. (keeps returning nil)